Refactor for future development

PeterKvayt opened the pull request:
Hi @umbracotrd!

As I can see you have merged my previous [pr](https://github.com/umbraco/Umbraco.Commerce.PaymentProviders.Worldpay/pull/6). But I did not confired it yet as you asked.

So here is new changes and you can fix problem with version.

Could you please increase minor version (to 10.1.0.0), so other developers should be aware to review changes before use new version.

Thanks.

PeterKvayt committed:
Fix for https://github.com/umbraco/Umbraco.Commerce.PaymentProviders.Worldpay/issues/4

PeterKvayt committed:
Refactoring

PeterKvayt committed:
Fix according to https://github.com/umbraco/Umbraco.Commerce.PaymentProviders.Worldpay/issues/5#issuecomment-2255314240

PeterKvayt committed:
Resolved according https://github.com/umbraco/Umbraco.Commerce.PaymentProviders.Worldpay/issues/7#issuecomment-2254969927

PeterKvayt committed:
Refactoring

PeterKvayt committed:
Pipeline fix

umbracotrd committed:
Update version.json

umbracotrd committed:
Corrected version

PeterKvayt committed:
Refactoring

PeterKvayt committed:
Merge branch 'temp/4' of https://github.com/PeterKvayt/Umbraco.Commerce.PaymentProviders.Worldpay into temp/4

PeterKvayt committed:
Merge branch 'support/10.x' into temp/4

# Conflicts:
#	src/Umbraco.Commerce.PaymentProviders.Worldpay/Constants/WorldpayValues.cs
#	src/Umbraco.Commerce.PaymentProviders.Worldpay/Extensions/PaymentStatusExtensions.cs
#	src/Umbraco.Commerce.PaymentProviders.Worldpay/WorldpayBusinessGateway350PaymentProvider.cs
#	src/Umbraco.Commerce.PaymentProviders.Worldpay/WorldpayBusinessGateway350Settings.cs

umbracotrd commented:
Hi @PeterKvayt , as I understand, this PR is just for refactoring code, right? I'm inclined to do this kind of things on v13 and above instead of v10. What do you think about this @mattbrailsford ?

PeterKvayt commented:
Hi @umbracotrd !

Yea, this is just refactoring, In some places log level/messages were changed for better log history.

mattbrailsford commented:
V13+ is cool with me. V10 is security fixes only now which this doesn’t look to fall into.

VictorTavkin commented:
Hi everyone!

@mattbrailsford, @umbracotrd, 

I would like to request your assistance in merging the proposed changes into v10. I completely understand that the version is currently in the support phase and it should not include any improvements and refactoring. @PeterKvayt and I have invested a significant amount of time in setting up the plugin, investigating/fixing the issues, and refactoring, but got the situation below. 

We’ve recently gone live with Umbraco v10 and encountered a few issues related to webhooks and logging, such as:
1) Receiving two callbacks from Worldpay for a single order: one from MC_callbackurl and another from PaymentResponse (requires investigation).
2) An error entry appears in the log files when a user cancels an order (the fix in the current PR).
3) Probably, there will be other issues in the future.

So, taking into account, the information above, now, we are in a situation where we can't further contribute to the plugin in v10, because the current PR is missing in that version.
Plus, we can't upgrade the website Umbraco version to v13+ to contribute to the version where the current PR is accepted.
So we can't develop either v10 or v13+. 

Additionally, given the current situation, two separate fixes are needed:
 - one for v10
 - another for v13+
This will take twice as long because a regular merge is not possible due to the absence of the current PR in v10

Thank you for your time and consideration.

umbracotrd commented:
Hi @VictorTavkin and @PeterKvayt , I appreciate your effort to contribute to this package. It's always nice to have more support from community.
Merging this PR for v10 seems fine for me, but it contains refactored code which makes it hard to cherry pick to v12, 13 and 14.
Let me bring this up to the team and get back to you.

umbracotrd committed:
Update version.json

umbracotrd commented:
Hi @VictorTavkin and @PeterKvayt , I've pushed version `10.1.0--preview.2+61ae4c6` to [Umbraco Nightly Build feed](https://docs.umbraco.com/umbraco-cms/fundamentals/setup/install/installing-nightly-builds). Could you confirm that it works as you expect?

We'll find a way to merge the changes up to newer versions later.

VictorTavkin commented:
Hi,
@mattbrailsford, @umbracotrd,
Thank you for your help. We'll return once we've tested the latest version.

umbracotrd commented:
Hi @VictorTavkin, how's it going with the 10.1.0 nightly build? Do you need more time to test it?

VictorTavkin commented:
Hi @umbracotrd, 
We plan to deploy the nightly build to our staging environment this week. then we'll get back to you as soon as we finish testing.

PeterKvayt commented:
@mattbrailsford
@umbracotrd 
@VictorTavkin 

I found issue in current pr, soon I will push fix.

PeterKvayt committed:
NullReferenceException fix.

In GetOrderReferenceAsync method context.Order is null, so we can't get OrderNumber from it.

PeterKvayt committed:
Merge branch 'temp/4' of https://github.com/PeterKvayt/Umbraco.Commerce.PaymentProviders.Worldpay into temp/4

PeterKvayt commented:
@umbracotrd 
@mattbrailsford 

Please, do not push new nightly build until I or @VictorTavkin do not ask about it.

PeterKvayt committed:
Ability to disable custom form fields added.

PeterKvayt commented:
[This commit](https://github.com/umbraco/Umbraco.Commerce.PaymentProviders.Worldpay/pull/8/commits/4b192691050a12a02aaed8ac6db4603a89945d5d) fixes NullreferenceException when verbosity logging is enabled.

[This commit](https://github.com/umbraco/Umbraco.Commerce.PaymentProviders.Worldpay/pull/8/commits/1137b37d0fe4f06789efd56476bf0e22436cc6c6) adds ability to disable custom form fields. It is can be useful for cases when you use `MC_callbackurl` instead of `Payment Response Url` (field in Worldpay management dashboard). 
We have conversation with Worldpay and they told us that we can use one of the folloing options to provide callback url: `MC_callbackurl` or `Payment Response Url`. To test this, we need to add ability to enable/disable `MC_callbackurl` option.

@umbracotrd @mattbrailsford  Could you please push new nightly build?

umbracotrd commented:
@PeterKvayt : I've pushed `10.1.0--preview.4+0603135` to Umbraco nightly feed

PeterKvayt commented:
> @PeterKvayt : I've pushed `10.1.0--preview.4+0603135` to Umbraco nightly feed

@umbracotrd Many thanks! We will aplly new `10.1.0--preview.4+0603135` as soon as possible.

PeterKvayt commented:
We already applied `10.1.0--preview.4+0603135` and now it is under testing.

VictorTavkin commented:
Hi!

Sorry for a delay with a feedback - the setup of production environment for Worldpay required some extra work and adjustments for our website plus a review of the setup from Worldpay's team. Therefore it took more time for setup and testing than it was initially expected.

Anyway, finally, we payed for an order using Worldpay payment method on production environment.

Many thanks for your patience and support! Please, merge the current PR. From our side no further changes are expected.

VictorTavkin commented:
Hi, @umbracotrd, @mattbrailsford!

Would you happen to have any updates regarding the current PR? We're looking forward to a stable build.

umbracotrd commented:
Sorry for not answering you sooner @VictorTavkin . It was a long holiday period. I'll merge the PR and release `10.1.0` in few hours

umbracotrd commented:
I've released `10.1.0`. Give it few minutes and you'll be able to search for it on nuget 🚀 . Thank you for your effort to contribute to this plugin.
